File: Day_1_Calorie_Counting/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// struct for maintaining top k integers
type TopKNums struct {
	K    int
	Nums []int
}

// initialize top k integers
func (tk *TopKNums) Init(k int) {
	tk.K = k
	tk.Nums = make([]int, tk.K)
	// initialize integers slice to zero
	for i := range tk.Nums {
		tk.Nums[i] = 0
	}
}

// insert new integer
func (tk *TopKNums) Insert(x int) {
	// return early if new integer is less than lowest value
	if x < tk.Nums[tk.K-1] {
		return
	}

	// perform linear scan to find where new integer should be inserted
	insert_idx := -1
	for i := range tk.Nums {
		if tk.Nums[i] > x {
			continue
		}

		insert_idx = i
		break
	}

	// shift slice to the right after insertion index
	copy(tk.Nums[insert_idx+1:tk.K], tk.Nums[insert_idx:tk.K-1])
	// insert new integer
	tk.Nums[insert_idx] = x
}

// compute top k sum
func (tk *TopKNums) Sum() int {
	sum := 0
	for i := range tk.Nums {
		sum += tk.Nums[i]
	}

	return sum
}

func main() {
	// file with input calories data
	filePath := "../calories.txt"
	// open and read file
	f, err := os.Open(filePath)
	if err != nil {
		fmt.Println("failed to open file", filePath, err.Error())
		return
	}

	// store top 3 calories
	k := 3
	topKCalories := TopKNums{}
	topKCalories.Init(k)
	// calorie count for current elf
	currentElfCalories := 0

	// scan input line by line
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()

		// check if blank line
		if len(strings.TrimSpace(line)) == 0 {
			// insert calorie count
			topKCalories.Insert(currentElfCalories)
			// set calorie count to zero for next elf
			currentElfCalories = 0
			continue
		}

		// convert calories string to int
		calories, err := strconv.Atoi(line)
		if err != nil {
			fmt.Println("failed to convert", line, "to integer", err.Error())
			return
		}

		// increment elf calorie count
		currentElfCalories += calories
	}

	// insert calorie count
	topKCalories.Insert(currentElfCalories)
	// compute and print sum of top 3 calories
	topKCaloriesSum := topKCalories.Sum()
	fmt.Println(topKCaloriesSum)
}

File: Day_1_Calorie_Counting/Rust/main.rs
use std::fs;
use std::collections::BinaryHeap;
use std::cmp::Reverse;

fn main() {
    // file with input calories data
    let file_path: &str = "../calories.txt";
    // read file contents to string
    let file_contents: String = fs::read_to_string(file_path).expect("failed to read file");

    // store top-k calories in min heap
    let k: i8 = 3;
    let mut top_k_calories: BinaryHeap<Reverse<u32>> = BinaryHeap::new();
    // initialize min heap with k zeros
    for _ in 0..k {
        top_k_calories.push(Reverse(0));
    }

    // loop over calories for each elf
    for elf_calories in file_contents.split("\n\n") {
        let mut elf_calories_sum: u32 = 0;
        // loop over each item for elf
        for calories in elf_calories.split("\n") {
            elf_calories_sum += calories.parse::<u32>().unwrap();
        }

        // push calories sum onto heap
        top_k_calories.push(Reverse(elf_calories_sum));
        // pop lowest calories off of heap
        top_k_calories.pop();
    }

    let mut top_k_calories_sum: u32 = 0;
    // compute sum over top-k calories
    for _ in 0..k {
        if let Some(Reverse(v)) = top_k_calories.pop() {
            top_k_calories_sum += v;
        }
    }

    println!("{top_k_calories_sum}");
}
